include/cuco/detail/bloom_filter/bloom_filter_impl.cuh

```diff
@@ -0,0 +1,359 @@
+
+  // TODO
+  // [[nodiscard]] __host__ double occupancy() const;
+  // [[nodiscard]] __host__ double expected_false_positive_rate(size_t unique_keys) const
+  // [[nodiscard]] __host__ __device__ static uint32_t optimal_pattern_bits(size_t num_blocks)
+  // template <typename CG, cuda::thread_scope NewScope = thread_scope>
+  // [[nodiscard]] __device__ constexpr auto make_copy(CG const& group, word_type* const
+  // memory_to_use, cuda_thread_scope<NewScope> scope = {}) const noexcept;
+
+ private:
+  template <class HashValue>
+  __device__ size_type block_idx(HashValue hash_value) const
+  {
+    // TODO use fast_int modulo
+    return hash_value % num_blocks_;
+  }
+
+  // we use the LSB bits of the hash value to determine the pattern bits for each word
+  template <class HashValue>
+  __device__ auto pattern(HashValue hash_value) const
+  {
+    cuda::std::array<word_type, block_words> pattern{};
+    auto constexpr word_bits           = sizeof(word_type) * CHAR_BIT;
+    auto constexpr bit_index_width     = cuda::std::bit_width(word_bits - 1);
+    word_type constexpr bit_index_mask = (word_type{1} << bit_index_width) - 1;
+
+    auto const bits_per_word = pattern_bits_ / block_words;
+    auto const remainder     = pattern_bits_ % block_words;
+
+    uint32_t k = 0;
+#pragma unroll block_words
+    for (int32_t i = 0; i < block_words; ++i) {
+      for (int32_t j = 0; j < bits_per_word + (i < remainder ? 1 : 0); ++j) {
+        if (k++ >= pattern_bits_) { return pattern; }
+        pattern[i] |= word_type{1} << (hash_value & bit_index_mask);
+        hash_value >>= bit_index_width;
+      }
+    }
+
+    return pattern;
+  }
+
+  template <class HashValue>
+  __device__ word_type pattern_word(HashValue hash_value, uint32_t i) const
+  {
+    auto constexpr word_bits           = sizeof(word_type) * CHAR_BIT;
+    auto constexpr bit_index_width     = cuda::std::bit_width(word_bits - 1);
+    word_type constexpr bit_index_mask = (word_type{1} << bit_index_width) - 1;
+
+    auto const bits_per_word = pattern_bits_ / block_words;
+    auto const remainder     = pattern_bits_ % block_words;
+    auto const bits_so_far   = bits_per_word * i + (i < remainder ? i : remainder);
+
+    hash_value >>= bits_so_far * bit_index_width;
+
+    // Compute the word at index i
+    word_type word  = 0;
+    int32_t j_limit = bits_per_word + (i < remainder ? 1 : 0);
+
+    for (int32_t j = 0; j < j_limit; ++j) {
+      word |= word_type{1} << (hash_value & bit_index_mask);
+      hash_value >>= bit_index_width;
+    }
+
+    return word;
+  }
+
+  template <uint32_t NumWords>
+  __device__ auto vec_load_words(size_type index) const
+  {
+    using vec_type = cuda::std::array<word_type, NumWords>;
+
+    return *reinterpret_cast<vec_type*>(
+      __builtin_assume_aligned(words_ + index, sizeof(word_type) * NumWords));
+  }
+
+  __host__ __device__ static constexpr size_t required_alignment() noexcept
+  {
+    return sizeof(word_type) * block_words;
```

Comment:
There should be an upper bound for the alignment, right? something like
```cpp
cuda::std::min(sizeof(word_type) * block_words, 16);
```

Reply:
I dunno about Nvidia GPU cache alignment issue, but on x86-64 host machines, you absolutely want blocks that are 32 bytes wide to be aligned at 32-byte boundaries - loads and stores go a lot faster that way. Same goes for 64-byte.

For larger block widths (not directly relevant here, but just FWIW) it's helpful not to span page boundaries for address lookup (rather than cache loading) effects, so large alignment can still be helpful for performance.

Reply:
> on x86-64 host machines, you absolutely want blocks that are 32 bytes wide to be aligned at 32-byte boundaries - loads and stores go a lot faster that way. Same goes for 64-byte.

Right, depending on the supported intrinsic sets like AVX2 (256-bit) or AVX512, modern x86 systems can benefit from 32B/64B alignment for optimal performance. GPUs, on the other hand, can fetch up to 128 bits in a single memory instruction for global memory.

Reply:
Implemented in [d2f04ff](https://github.com/NVIDIA/cuCollections/pull/573/commits/d2f04ff2e95bf04c4f0db2c3f6f3a97c744e5c4d)